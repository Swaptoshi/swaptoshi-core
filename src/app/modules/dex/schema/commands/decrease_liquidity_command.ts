export const decreaseLiquidityCommandSchema = {
	$id: '/dex/command/decrease_liquidity',
	type: 'object',
	properties: {
		poolAddress: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
		tokenId: {
			dataType: 'string',
			fieldNumber: 2,
		},
		liquidity: {
			dataType: 'string',
			fieldNumber: 3,
		},
		amount0Min: {
			dataType: 'string',
			fieldNumber: 4,
		},
		amount1Min: {
			dataType: 'string',
			fieldNumber: 5,
		},
		deadline: {
			dataType: 'string',
			fieldNumber: 6,
		},
	},
};
