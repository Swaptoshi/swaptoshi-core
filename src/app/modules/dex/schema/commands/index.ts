export * from './burn_command';
export * from './collect_command';
export * from './collect_treasury_command';
export * from './create_pool_command';
export * from './decrease_liquidity_command';
export * from './exact_input_command';
export * from './exact_input_single_command';
export * from './exact_output_command';
export * from './exact_output_single_command';
export * from './increase_liquidity_command';
export * from './mint_command';
export * from './treasurify_command';
