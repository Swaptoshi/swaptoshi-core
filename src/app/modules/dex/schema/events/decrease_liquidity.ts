export const decreaseLiquidityEventSchema = {
	$id: '/dex/events/decrease_liquidity',
	type: 'object',
	required: ['tokenId', 'liquidity', 'amount0', 'amount1'],
	properties: {
		tokenId: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
		liquidity: {
			dataType: 'string',
			fieldNumber: 2,
		},
		amount0: {
			dataType: 'string',
			fieldNumber: 3,
		},
		amount1: {
			dataType: 'string',
			fieldNumber: 4,
		},
	},
};
