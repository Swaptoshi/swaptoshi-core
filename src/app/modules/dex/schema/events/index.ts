export * from './burn';
export * from './collect';
export * from './collect_position';
export * from './collect_protocol';
export * from './decrease_liquidity';
export * from './flash';
export * from './increase_liquidity';
export * from './increase_observation_cardinality_next';
export * from './mint';
export * from './pool_created';
export * from './pool_initialized';
export * from './swap';
export * from './token_registered';
export * from './tokenuri_created';
export * from './tokenuri_destroyed';
export * from './treasurify';
