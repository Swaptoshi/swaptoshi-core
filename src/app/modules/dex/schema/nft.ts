const LENGTH_NFT_ID = 16;
const MIN_LENGTH_MODULE_NAME = 1;
const MAX_LENGTH_MODULE_NAME = 32;
const LENGTH_COLLECTION_ID = 4;

export const genesisNFTStoreSchema = {
	$id: '/nft/module/genesis',
	type: 'object',
	required: ['nftSubstore', 'supportedNFTsSubstore'],
	properties: {
		nftSubstore: {
			type: 'array',
			fieldNumber: 1,
			items: {
				type: 'object',
				required: ['nftID', 'owner', 'attributesArray'],
				properties: {
					nftID: {
						dataType: 'bytes',
						minLength: LENGTH_NFT_ID,
						maxLength: LENGTH_NFT_ID,
						fieldNumber: 1,
					},
					owner: {
						dataType: 'bytes',
						fieldNumber: 2,
					},
					attributesArray: {
						type: 'array',
						fieldNumber: 3,
						items: {
							type: 'object',
							required: ['module', 'attributes'],
							properties: {
								module: {
									dataType: 'string',
									minLength: MIN_LENGTH_MODULE_NAME,
									maxLength: MAX_LENGTH_MODULE_NAME,
									pattern: '^[a-zA-Z0-9]*$',
									fieldNumber: 1,
								},
								attributes: {
									dataType: 'bytes',
									fieldNumber: 2,
								},
							},
						},
					},
				},
			},
		},
		supportedNFTsSubstore: {
			type: 'array',
			fieldNumber: 2,
			items: {
				type: 'object',
				required: ['chainID', 'supportedCollectionIDArray'],
				properties: {
					chainID: {
						dataType: 'bytes',
						fieldNumber: 1,
					},
					supportedCollectionIDArray: {
						type: 'array',
						fieldNumber: 2,
						items: {
							type: 'object',
							required: ['collectionID'],
							properties: {
								collectionID: {
									dataType: 'bytes',
									minLength: LENGTH_COLLECTION_ID,
									maxLength: LENGTH_COLLECTION_ID,
									fieldNumber: 1,
								},
							},
						},
					},
				},
			},
		},
	},
};
