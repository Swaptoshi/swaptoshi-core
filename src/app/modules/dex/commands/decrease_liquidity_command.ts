/* eslint-disable class-methods-use-this */

import { BaseCommand, CommandVerifyContext, CommandExecuteContext, VerificationResult, VerifyStatus } from 'klayr-sdk';
import { PositionManagerStore } from '../stores/position_manager';
import { commandSwapContext } from '../stores/context';
import { decreaseLiquidityCommandSchema } from '../schema';
import { DecreaseLiquidityParams } from '../types';
import { verifyDecreaseLiquidityParam } from '../utils';

export class DecreaseLiquidityCommand extends BaseCommand {
	// eslint-disable-next-line @typescript-eslint/require-await
	public async verify(_context: CommandVerifyContext<DecreaseLiquidityParams>): Promise<VerificationResult> {
		try {
			verifyDecreaseLiquidityParam(_context.params);
		} catch (error: unknown) {
			return {
				status: VerifyStatus.FAIL,
				error: new Error((error as { message: string }).message),
			};
		}
		return { status: VerifyStatus.OK };
	}

	public async execute(_context: CommandExecuteContext<DecreaseLiquidityParams>): Promise<void> {
		const positionManagerStore = this.stores.get(PositionManagerStore);
		const context = commandSwapContext(_context);
		const positionManager = await positionManagerStore.getMutablePositionManager(context, _context.params.poolAddress);
		await positionManager.decreaseLiquidity(_context.params);
	}

	public schema = decreaseLiquidityCommandSchema;
}
