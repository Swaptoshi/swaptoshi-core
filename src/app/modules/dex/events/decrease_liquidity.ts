/* eslint-disable import/no-cycle */
import { Modules } from 'klayr-sdk';
import { decreaseLiquidityEventSchema } from '../schema';

export interface DecreaseLiquidityEventData {
	tokenId: Buffer;
	liquidity: string;
	amount0: string;
	amount1: string;
}

export class DecreaseLiquidityEvent extends Modules.BaseEvent<DecreaseLiquidityEventData> {
	public schema = decreaseLiquidityEventSchema;
}
