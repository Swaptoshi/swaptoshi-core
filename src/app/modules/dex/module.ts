/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable class-methods-use-this */
/* eslint-disable @typescript-eslint/member-ordering */

import {
	BaseInteroperableModule,
	BlockExecuteContext,
	FeeMethod,
	MainchainInteroperabilityMethod,
	ModuleInitArgs,
	ModuleMetadata,
	NFTMethod,
	SidechainInteroperabilityMethod,
	TokenMethod,
	TransactionExecuteContext,
	TransactionVerifyContext,
	VerificationResult,
	VerifyStatus,
} from 'lisk-sdk';
import { BurnCommand } from './commands/burn_command';
import { CollectCommand } from './commands/collect_command';
import { CreatePoolCommand } from './commands/create_pool_command';
import { DecreaseLiquidityCommand } from './commands/decrease_liquidity_command';
import { ExactInputCommand } from './commands/exact_input_command';
import { ExactInputSingleCommand } from './commands/exact_input_single_command';
import { ExactOutputCommand } from './commands/exact_output_command';
import { ExactOutputSingleCommand } from './commands/exact_output_single_command';
import { IncreaseLiquidityCommand } from './commands/increase_liquidity_command';
import { MintCommand } from './commands/mint_command';
import { TreasurifyCommand } from './commands/treasurify_command';
import { DexEndpoint } from './endpoint';
import { BurnEvent } from './events/burn';
import { CollectEvent } from './events/collect';
import { CollectPositionEvent } from './events/collect_position';
import { CollectProtocolEvent } from './events/collect_protocol';
import { DecreaseLiquidityEvent } from './events/decrease_liquidity';
import { FlashEvent } from './events/flash';
import { IncreaseLiquidityEvent } from './events/increase_liquidity';
import { IncreaseObservationCardinalityNextEvent } from './events/increase_observation_cardinality_next';
import { MintEvent } from './events/mint';
import { PoolCreatedEvent } from './events/pool_created';
import { PoolInitializedEvent } from './events/pool_initialized';
import { SwapEvent } from './events/swap';
import { TokenURICreatedEvent } from './events/tokenuri_created';
import { TokenURIDestroyedEvent } from './events/tokenuri_destroyed';
import { DexMethod } from './method';
import {
	getPoolEndpointRequestSchema,
	getPoolEndpointResponseSchema,
} from './schema/endpoint/get_pool';
import {
	getPositionEndpointRequestSchema,
	getPositionEndpointResponseSchema,
} from './schema/endpoint/get_position';
import {
	getTokenURIEndpointRequestSchema,
	getTokenURIEndpointResponseSchema,
} from './schema/endpoint/get_token_uri';
import {
	observeEndpointRequestSchema,
	observeEndpointResponseSchema,
} from './schema/endpoint/observe';
import {
	quoteExactInputEndpointRequestSchema,
	quoteExactInputEndpointResponseSchema,
} from './schema/endpoint/quote_exact_input';
import {
	quoteExactInputSingleEndpointRequestSchema,
	quoteExactInputSingleEndpointResponseSchema,
} from './schema/endpoint/quote_exact_input_single';
import {
	quoteExactOutputEndpointRequestSchema,
	quoteExactOutputEndpointResponseSchema,
} from './schema/endpoint/quote_exact_output';
import {
	quoteExactOutputSingleEndpointRequestSchema,
	quoteExactOutputSingleEndpointResponseSchema,
} from './schema/endpoint/quote_exact_output_single';
import { ObservationStore } from './stores/observation';
import { PoolStore } from './stores/pool';
import { PositionInfoStore } from './stores/position_info';
import { PositionManagerStore } from './stores/position_manager';
import { TickBitmapStore } from './stores/tick_bitmap';
import { TickInfoStore } from './stores/tick_info';
import { TokenSymbolStore } from './stores/token_symbol';
import { DexModuleConfig } from './types';
import { verifyMinimumFee, verifySwapByTransfer, executeSwapByTransfer } from './hooks';
import { defaultConfig } from './constants';
import {
	getMetadataEndpointRequestSchema,
	getMetadataEndpointResponseSchema,
} from './schema/endpoint/get_metadata';
import { verifyFeeConversion } from './hooks/verifyFeeConversion';
import { executeFeeConversion } from './hooks/executeFeeConversion';
import { TreasurifyEvent } from './events/treasurify';
import {
	getPoolAddressFromCollectionIdEndpointRequestSchema,
	getPoolAddressFromCollectionIdEndpointResponseSchema,
} from './schema/endpoint/get_pool_address_from_collection_id';
import {
	quotePriceEndpointRequestSchema,
	quotePriceEndpointResponseSchema,
} from './schema/endpoint/quote_price';
import { TokenRegisteredEvent } from './events/token_registered';
import { SupportedTokenStore } from './stores/supported_token';
import { verifyValidTransfer } from './hooks/verifyValidTransfer';
import { DexInteroperableMethod } from './cc_method';

export class DexModule extends BaseInteroperableModule {
	public _config: DexModuleConfig | undefined;
	public _feeMethod: FeeMethod | undefined;
	public _tokenMethod: TokenMethod | undefined;
	public _dexInteroperableMethod = new DexInteroperableMethod(this.stores, this.events);

	public crossChainCommand = [];
	public crossChainMethod = this._dexInteroperableMethod;

	public endpoint = new DexEndpoint(this.stores, this.offchainStores);
	public method = new DexMethod(this.stores, this.events);
	public commands = [
		new CreatePoolCommand(this.stores, this.events),
		new MintCommand(this.stores, this.events),
		new IncreaseLiquidityCommand(this.stores, this.events),
		new DecreaseLiquidityCommand(this.stores, this.events),
		new CollectCommand(this.stores, this.events),
		new BurnCommand(this.stores, this.events),
		new ExactInputCommand(this.stores, this.events),
		new ExactInputSingleCommand(this.stores, this.events),
		new ExactOutputCommand(this.stores, this.events),
		new ExactOutputSingleCommand(this.stores, this.events),
		new TreasurifyCommand(this.stores, this.events),
	];

	public constructor() {
		super();
		// registeration of stores and events
		this.stores.register(PoolStore, new PoolStore(this.name, 0, this.stores, this.events));
		this.stores.register(
			PositionManagerStore,
			new PositionManagerStore(this.name, 1, this.stores, this.events),
		);

		this.stores.register(ObservationStore, new ObservationStore(this.name, 2));
		this.stores.register(PositionInfoStore, new PositionInfoStore(this.name, 3));
		this.stores.register(TickBitmapStore, new TickBitmapStore(this.name, 4));
		this.stores.register(TickInfoStore, new TickInfoStore(this.name, 5));
		this.stores.register(TokenSymbolStore, new TokenSymbolStore(this.name, 6, this.events));
		this.stores.register(SupportedTokenStore, new SupportedTokenStore(this.name, 7));

		this.events.register(BurnEvent, new BurnEvent(this.name));
		this.events.register(CollectPositionEvent, new CollectPositionEvent(this.name));
		this.events.register(CollectProtocolEvent, new CollectProtocolEvent(this.name));
		this.events.register(CollectEvent, new CollectEvent(this.name));
		this.events.register(DecreaseLiquidityEvent, new DecreaseLiquidityEvent(this.name));
		this.events.register(FlashEvent, new FlashEvent(this.name));
		this.events.register(IncreaseLiquidityEvent, new IncreaseLiquidityEvent(this.name));
		this.events.register(
			IncreaseObservationCardinalityNextEvent,
			new IncreaseObservationCardinalityNextEvent(this.name),
		);
		this.events.register(MintEvent, new MintEvent(this.name));
		this.events.register(PoolCreatedEvent, new PoolCreatedEvent(this.name));
		this.events.register(PoolInitializedEvent, new PoolInitializedEvent(this.name));
		this.events.register(SwapEvent, new SwapEvent(this.name));
		this.events.register(TokenURICreatedEvent, new TokenURICreatedEvent(this.name));
		this.events.register(TokenURIDestroyedEvent, new TokenURIDestroyedEvent(this.name));
		this.events.register(TreasurifyEvent, new TreasurifyEvent(this.name));
		this.events.register(TokenRegisteredEvent, new TokenRegisteredEvent(this.name));
	}

	public addDependencies(
		tokenMethod: TokenMethod,
		nftMethod: NFTMethod,
		feeMethod: FeeMethod,
		interoperabilityMethod: SidechainInteroperabilityMethod | MainchainInteroperabilityMethod,
	) {
		const poolStore = this.stores.get(PoolStore);
		const positionManagerStore = this.stores.get(PositionManagerStore);
		const supportManagerStore = this.stores.get(SupportedTokenStore);

		poolStore.addDependencies(tokenMethod);
		positionManagerStore.addDependencies(tokenMethod, nftMethod);
		supportManagerStore.addDependencies(tokenMethod);

		this._feeMethod = feeMethod;
		this._tokenMethod = tokenMethod;
		this._dexInteroperableMethod.addDependencies(interoperabilityMethod, tokenMethod, nftMethod);
	}

	public metadata(): ModuleMetadata {
		return {
			...this.baseMetadata(),
			endpoints: [
				{
					name: this.endpoint.getPool.name,
					request: getPoolEndpointRequestSchema,
					response: getPoolEndpointResponseSchema,
				},
				{
					name: this.endpoint.getPosition.name,
					request: getPositionEndpointRequestSchema,
					response: getPositionEndpointResponseSchema,
				},
				{
					name: this.endpoint.getTokenURI.name,
					request: getTokenURIEndpointRequestSchema,
					response: getTokenURIEndpointResponseSchema,
				},
				{
					name: this.endpoint.getPoolAddressFromCollectionId.name,
					request: getPoolAddressFromCollectionIdEndpointRequestSchema,
					response: getPoolAddressFromCollectionIdEndpointResponseSchema,
				},
				{
					name: this.endpoint.getMetadata.name,
					request: getMetadataEndpointRequestSchema,
					response: getMetadataEndpointResponseSchema,
				},
				{
					name: this.endpoint.observe.name,
					request: observeEndpointRequestSchema,
					response: observeEndpointResponseSchema,
				},
				{
					name: this.endpoint.quoteExactInput.name,
					request: quoteExactInputEndpointRequestSchema,
					response: quoteExactInputEndpointResponseSchema,
				},
				{
					name: this.endpoint.quoteExactInputSingle.name,
					request: quoteExactInputSingleEndpointRequestSchema,
					response: quoteExactInputSingleEndpointResponseSchema,
				},
				{
					name: this.endpoint.quoteExactOutput.name,
					request: quoteExactOutputEndpointRequestSchema,
					response: quoteExactOutputEndpointResponseSchema,
				},
				{
					name: this.endpoint.quoteExactOutputSingle.name,
					request: quoteExactOutputSingleEndpointRequestSchema,
					response: quoteExactOutputSingleEndpointResponseSchema,
				},
				{
					name: this.endpoint.quotePrice.name,
					request: quotePriceEndpointRequestSchema,
					response: quotePriceEndpointResponseSchema,
				},
			],
			assets: [],
		};
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async init(_args: ModuleInitArgs): Promise<void> {
		this._config = { ...defaultConfig, ..._args.moduleConfig };
		const poolStore = this.stores.get(PoolStore);
		const positionManagerStore = this.stores.get(PositionManagerStore);
		const supportedTokenStore = this.stores.get(SupportedTokenStore);

		poolStore.init(this._config);
		positionManagerStore.init(Buffer.from(_args.genesisConfig.chainID, 'hex'));
		supportedTokenStore.init(this._config);
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async verifyTransaction(_context: TransactionVerifyContext): Promise<VerificationResult> {
		try {
			await verifyMinimumFee(_context, this._config!);
			await verifyValidTransfer.bind(this)(_context);
			await verifySwapByTransfer.bind(this)(_context);
			await verifyFeeConversion.bind(this)(_context);
		} catch (error: unknown) {
			return {
				status: VerifyStatus.FAIL,
				error: new Error((error as { message: string }).message),
			};
		}
		return { status: VerifyStatus.OK };
	}

	public async beforeCommandExecute(_context: TransactionExecuteContext): Promise<void> {
		await verifyMinimumFee(_context, this._config!);
		await verifyValidTransfer.bind(this)(_context);
		await verifySwapByTransfer.bind(this)(_context);
		await executeFeeConversion.bind(this)(_context);
	}

	public async afterCommandExecute(_context: TransactionExecuteContext): Promise<void> {
		await executeSwapByTransfer.bind(this)(_context);
	}

	public async beforeTransactionsExecute(context: BlockExecuteContext): Promise<void> {
		await this.stores.get(SupportedTokenStore).apply(context);
	}
}
