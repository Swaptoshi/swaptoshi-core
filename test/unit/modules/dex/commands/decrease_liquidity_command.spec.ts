/* eslint-disable jest/expect-expect */
import { CommandExecuteContext, CommandVerifyContext, VerifyStatus } from 'lisk-sdk';
import { DexModule } from '../../../../../src/app/modules/dex/module';
import { DecreaseLiquidityParams } from '../../../../../src/app/modules/dex/types';
import { invalidAddress, invalidNumberString } from '../utils/invalid';
import { Tokens, commandFixture } from '../utils/fixtures';
import { NonfungiblePositionManager } from '../../../../../src/app/modules/dex/stores/factory';
import { FeeAmount, TICK_SPACINGS, getMaxTick, getMinTick } from '../stores/shared/utilities';
import { poolAddress, senderAddress, senderPublicKey } from '../utils/account';
import { eventResultHaveMinimumLength } from '../../../../utils/events';
import { NFTRegistry } from '../stores/shared/nft/nft_registry';
import { TokenRegistry } from '../stores/shared/token/token_registry';
import { decreaseLiquidityCommandSchema } from '../../../../../src/app/modules/dex/schema/commands/decrease_liquidity_command';
import { DecreaseLiquidityCommand } from '../../../../../src/app/modules/dex/commands/decrease_liquidity_command';
import { DecreaseLiquidityEvent } from '../../../../../src/app/modules/dex/events/decrease_liquidity';

type CommandParam = DecreaseLiquidityParams;
const COMMAND_NAME = 'decreaseLiquidity';
const commandSchema = decreaseLiquidityCommandSchema;

const validParam: CommandParam = {
	poolAddress,
	tokenId: '0',
	liquidity: '25',
	amount0Min: '0',
	amount1Min: '0',
	deadline: Date.now().toString(),
};

describe('DecreaseLiquidityCommand', () => {
	let module: DexModule;
	let command: DecreaseLiquidityCommand;
	let nft: NonfungiblePositionManager;
	let tokens: Tokens;
	let createCommandVerifyContext: (params: CommandParam) => CommandVerifyContext<CommandParam>;
	let createCommandExecuteContext: (params: CommandParam) => CommandExecuteContext<CommandParam>;

	beforeEach(async () => {
		({ module, createCommandExecuteContext, createCommandVerifyContext, tokens, nft } =
			await commandFixture<CommandParam>(COMMAND_NAME, commandSchema, senderPublicKey, validParam));
		command = new DecreaseLiquidityCommand(module.stores, module.events);

		await nft.mint({
			token0: tokens[0].address,
			token1: tokens[1].address,
			fee: FeeAmount.MEDIUM,
			tickLower: getMinTick(TICK_SPACINGS[FeeAmount.MEDIUM]).toString(),
			tickUpper: getMaxTick(TICK_SPACINGS[FeeAmount.MEDIUM]).toString(),
			recipient: senderAddress,
			amount0Desired: '100',
			amount1Desired: '100',
			amount0Min: '0',
			amount1Min: '0',
			deadline: Date.now().toString(),
		});
	});

	afterEach(() => {
		NFTRegistry.reset();
		TokenRegistry.reset();
	});

	describe('constructor', () => {
		it('should have valid name', () => {
			expect(command.name).toBe(COMMAND_NAME);
		});

		it('should have valid schema', () => {
			expect(command.schema).toMatchSnapshot();
		});
	});

	describe('verify', () => {
		it('should return status OK when called with valid input', async () => {
			const context = createCommandVerifyContext(validParam);
			await expect(command.verify(context)).resolves.toHaveProperty('status', VerifyStatus.OK);
		});

		it('should throw error when user sends transaction with invalid address (poolAddress)', async () => {
			const context = createCommandVerifyContext({
				...validParam,
				poolAddress: invalidAddress,
			});
			await expect(command.verify(context)).resolves.toHaveProperty('status', VerifyStatus.FAIL);
		});

		it('should throw error when user sends transaction with invalid number string (tokenId)', async () => {
			const context = createCommandVerifyContext({
				...validParam,
				tokenId: invalidNumberString,
			});
			await expect(command.verify(context)).resolves.toHaveProperty('status', VerifyStatus.FAIL);
		});

		it('should throw error when user sends transaction with invalid number string (liquidity)', async () => {
			const context = createCommandVerifyContext({
				...validParam,
				liquidity: invalidNumberString,
			});
			await expect(command.verify(context)).resolves.toHaveProperty('status', VerifyStatus.FAIL);
		});

		it('should throw error when user sends transaction with invalid number string (amount0Min)', async () => {
			const context = createCommandVerifyContext({
				...validParam,
				amount0Min: invalidNumberString,
			});
			await expect(command.verify(context)).resolves.toHaveProperty('status', VerifyStatus.FAIL);
		});

		it('should throw error when user sends transaction with invalid number string (amount1Min)', async () => {
			const context = createCommandVerifyContext({
				...validParam,
				amount1Min: invalidNumberString,
			});
			await expect(command.verify(context)).resolves.toHaveProperty('status', VerifyStatus.FAIL);
		});

		it('should throw error when user sends transaction with invalid number string (deadline)', async () => {
			const context = createCommandVerifyContext({
				...validParam,
				deadline: invalidNumberString,
			});
			await expect(command.verify(context)).resolves.toHaveProperty('status', VerifyStatus.FAIL);
		});
	});

	describe('execute', () => {
		it('should decreases position liquidity', async () => {
			const context = createCommandExecuteContext(validParam);
			await command.execute(context);

			const { liquidity } = await nft.getPositions(validParam.tokenId);
			expect(liquidity).toBe('75');
		});

		it('should add command events', async () => {
			const context = createCommandExecuteContext(validParam);
			await command.execute(context);

			eventResultHaveMinimumLength(context.eventQueue, DecreaseLiquidityEvent, module.name, 1);
		});
	});
});
