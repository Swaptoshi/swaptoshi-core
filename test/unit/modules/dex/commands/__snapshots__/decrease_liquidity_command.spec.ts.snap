// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DecreaseLiquidityCommand constructor should have valid schema 1`] = `
{
  "$id": "/dex/command/decrease_liquidity",
  "properties": {
    "amount0Min": {
      "dataType": "string",
      "fieldNumber": 4,
    },
    "amount1Min": {
      "dataType": "string",
      "fieldNumber": 5,
    },
    "deadline": {
      "dataType": "string",
      "fieldNumber": 6,
    },
    "liquidity": {
      "dataType": "string",
      "fieldNumber": 3,
    },
    "poolAddress": {
      "dataType": "bytes",
      "fieldNumber": 1,
      "format": "klayr32",
    },
    "tokenId": {
      "dataType": "string",
      "fieldNumber": 2,
    },
  },
  "type": "object",
}
`;
